mozilla_version/test/test_fenix.py

```diff
@@ -0,0 +1,307 @@
+
+from mozilla_version.errors import PatternNotMatchedError, TooManyTypesError, NoVersionTypeError
+from mozilla_version.fenix import FenixVersion
+from mozilla_version.test import does_not_raise
+
+
+VALID_VERSIONS = {
+    '0.3.0-rc.1': 'release_candidate',
+    '1.0.0': 'release',
+    '1.0.0-rc.1': 'release_candidate',
+    '1.0.0-rc.2': 'release_candidate',
+    '1.0.1': 'release',
+    '1.0.1-rc.1': 'release_candidate',
+    '1.1.0': 'release',
+    '1.1.0-rc.1': 'release_candidate',
+    '1.1.0-rc.2': 'release_candidate',
+    '1.2.0': 'release',
+    '1.2.0-rc.1': 'release_candidate',
+    '1.2.0-rc.2': 'release_candidate',
+    '1.3.0': 'release',
+    '1.3.0-rc.1': 'release_candidate',
+    '1.3.0-rc.2': 'release_candidate',
+    '1.3.0-rc.3': 'release_candidate',
+    '2.2.0-rc.1': 'release_candidate',
+    '3.0.0-beta.2': 'beta',
+    '3.0.0-beta.3': 'beta',
+    '3.0.0': 'release',
+}
+
+
+@pytest.mark.parametrize('major_number, minor_number, patch_number, beta_number, rc_number, expected_output_string', ((
+    3, 0, None, None, None, '3.0'
+), (
+    3, 0, 1, None, None, '3.0.1'
+), (
+    3, 0, None, 3, None, '3.0-beta.3'
+), (
+    3, 0, None, None, 3, '3.0-rc.3'
+)))
+def test_fenix_version_constructor_and_str(major_number, minor_number, patch_number, beta_number, rc_number, expected_output_string):
+    assert str(FenixVersion(
+        major_number=major_number,
+        minor_number=minor_number,
+        patch_number=patch_number,
+        beta_number=beta_number,
+        release_candidate_number=rc_number,
+    )) == expected_output_string
+
+
+@pytest.mark.parametrize('major_number, minor_number, patch_number, beta_number, rc_number, ExpectedErrorType', ((
+    3, 0, None, 1, 1, TooManyTypesError
+), (
+    3, 0, None, 0, None, ValueError
+), (
+    3, 0, None, None, 0, ValueError
+), (
+    -1, 0, None, None, None, ValueError
+), (
+    3, -1, None, None, None, ValueError
+), (
+    3, 0, -1, None, None, ValueError
+), (
+    2.2, 0, 0, None, None, ValueError
+), (
+    'some string', 0, 0, None, None, ValueError
+)))
+def test_fail_fenix_version_constructor(major_number, minor_number, patch_number, beta_number, rc_number, ExpectedErrorType):
+    with pytest.raises(ExpectedErrorType):
+        FenixVersion(
+            major_number=major_number,
+            minor_number=minor_number,
+            patch_number=patch_number,
+            beta_number=beta_number,
+            release_candidate_number=rc_number,
+        )
+
+
+def test_fenix_version_constructor_minimum_kwargs():
+    assert str(FenixVersion(3, 0)) == '3.0'
+    assert str(FenixVersion(3, 0, 1)) == '3.0.1'
+    assert str(FenixVersion(3, 1, 0)) == '3.1.0'
+    assert str(FenixVersion(3, 0, beta_number=1)) == '3.0-beta.1'
+
+    assert str(FenixVersion(1, 0, release_candidate_number=1)) == '1.0-rc.1'
+
+
+@pytest.mark.parametrize('version_string, ExpectedErrorType', (
+    ('1.0.0b1', PatternNotMatchedError),
+    ('1.0.0.0b1', ValueError),
```

Comment:
Question: I've been wondering if throwing 2 different type of errors (`PatternNotMatchedError` and `ValueError`) makes sense. Do you think `parse()` should throw only a single error, one day? 

(No action needed in this PR)

Reply:
Hrm, this is an interesting question ... not sure

Reply:
No problem. We'll figure this out along the way.